buildSrc/src/main/java/com/google/firebase/gradle/ReleaseGenerator.kt

```diff
@@ -0,0 +1,118 @@
+
+open class ReleaseGenerator : DefaultTask() {
+    @TaskAction
+    @Throws(Exception::class)
+    fun generateReleaseConfig() {
+        val currentRelease = project.property("currentRelease").toString()
+        val pastRelease = project.property("pastRelease").toString()
+        val rootDir = project.rootDir
+        val availableModules = parseSubProjects(rootDir)
+
+        val repo = initRepo(rootDir)
+        val headRef = repo.repository.resolve(Constants.HEAD)
+        val branchRef = getObjectRefForBranchName(repo, pastRelease)
+
+        val changedDirs = getChangedLibraries(repo, branchRef, headRef, FIREBASE_LIBRARIES)
+        val changedModules = findChangedModules(changedDirs, availableModules)
+        writeReleaseConfig(rootDir, changedModules, currentRelease)
+    }
+
+    private fun findChangedModules(changedDirs: List<String>, availableModules: Set<String>): List<String> = changedDirs.flatMap { directory ->
+        val moduleName = directory.replace("/", ":")
+        val ktxModuleName = "$moduleName:ktx"
+        listOf(moduleName, ktxModuleName).filter { availableModules.contains(it) }
+    }
+
+    private fun parseSubProjects(rootDir: File) = File(rootDir, "subprojects.cfg").readLines().filter { !it.startsWith("#") }.toSet()
+
+    @Throws(IOException::class)
+    private fun initRepo(repoPath: File): Git {
+        if (!repoPath.exists()) {
+            throw RuntimeException("Repo path doesn't exist")
+        }
+        return Git.open(repoPath)
+    }
+
+    @Throws(GitAPIException::class)
+    private fun getObjectRefForBranchName(repo: Git, branchName: String): ObjectId {
+        return repo.branchList().setListMode(ListBranchCommand.ListMode.REMOTE).call()
+                .filter { ref: Ref -> ref.name == "refs/remotes/origin/$branchName" }
+                .map { obj: Ref -> obj.objectId }.firstOrNull()
+                ?: throw RuntimeException("Could not find branch named $branchName")
+    }
+
+    private fun getChangedLibraries(repo: Git, startRef: ObjectId, endRef: ObjectId, directories: List<String>): List<String> {
+        return FIREBASE_LIBRARIES.filter { libraryPath: String ->
+            repo.log()
+                    .addPath("$libraryPath/*.kt")
+                    .addPath("$libraryPath/*.java")
+                    .addPath("$libraryPath/CHANGELOG.md")
+                    .addPath("$libraryPath/*.gradle")
+                    .addRange(startRef, endRef)
+                    .setMaxCount(1)
+                    .call()
+                    .iterator().hasNext()
+        }
+    }
+
+    private fun writeReleaseConfig(configPath: File, libraries: List<String>, releaseName: String) {
+        File(configPath, "release.cfg").writeText(
+                """
+                    [release]
+                    name = $releaseName
+                    mode = RELEASE
+                    
+                    [modules]
+                    ${libraries.joinToString("\n                    ")}
```

Comment:
Theres a couple ways you could avoid needing that long string for padding that would be more type safe:

```kotlin
${libraries.joinToString { "$it\n".padEnd(20, ' ') }}
```
```kotlin
${libraries.joinToString("\n".padEnd(20, ' '))}
```
```kotlin
${libraries.joinToString("\n", postfix = " ".repeat(20))}
```